src/libexpr/include/nix/expr/value.hh

```diff
@@ -447,21 +513,42 @@ class ValueStorage<ptrSize, std::enable_if_t<detail::useBitPackedValueStorage<pt
 
     PrimaryDiscriminator getPrimaryDiscriminator() const noexcept
     {
-        return static_cast<PrimaryDiscriminator>(payload[0] & discriminatorMask);
+        return static_cast<PrimaryDiscriminator>(p0 & discriminatorMask);
     }
 
     static void assertAligned(PackedPointer val) noexcept
     {
         assert((val & discriminatorMask) == 0 && "Pointer is not 8 bytes aligned");
     }
 
+    void finish(PackedPointer p0_, PackedPointer p1_)
+    {
+        debug("FINISH %x %08x %08x", this, p0_, p1_);
+
+        // Note: p1 *must* be updated before p0.
+        p1 = p1_;
```

Comment:
Maybe see: my previous 2 comments on using a fence + abstracted function to try to enforce this?